perl/lib/Nix/Store.xs

```diff
@@ -17,100 +17,143 @@
 #include <sodium.h>
 #include <nlohmann/json.hpp>
 
-
 using namespace nix;
 
+static bool libStoreInitialized = false;
 
-static ref<Store> store()
-{
-    static std::shared_ptr<Store> _store;
-    if (!_store) {
-        try {
-            initLibStore();
-            _store = openStore();
-        } catch (Error & e) {
-            croak("%s", e.what());
-        }
-    }
-    return ref<Store>(_store);
-}
-
+struct StoreWrapper {
+    ref<Store> store;
+};
 
 MODULE = Nix::Store PACKAGE = Nix::Store
 PROTOTYPES: ENABLE
 
+TYPEMAP: <<HERE
+StoreWrapper *      O_OBJECT
+
+OUTPUT
+O_OBJECT
+    sv_setref_pv( $arg, CLASS, (void*)$var );
+
+INPUT
+O_OBJECT
+    if ( sv_isobject($arg) && (SvTYPE(SvRV($arg)) == SVt_PVMG) ) {
+        $var = ($type)SvIV((SV*)SvRV( $arg ));
+    }
+    else {
+        warn( \"${Package}::$func_name() -- \"
+		\"$var not a blessed SV reference\");
+        XSRETURN_UNDEF;
+    }
+HERE
 
 #undef dNOOP // Hack to work around "error: declaration of 'Perl___notused' has a different language linkage" error message on clang.
 #define dNOOP
 
+void
+StoreWrapper::DESTROY()
+
+StoreWrapper *
+StoreWrapper::new(char * s = nullptr)
+    CODE:
+        static std::shared_ptr<Store> _store;
+        try {
+            if (!libStoreInitialized) {
+                initLibStore();
```

Comment:
`initLibStore()` internally already has a `initLibStoreDone` flag, so `libStoreInitialized` seems superfluous.

Reply:
That seems to be for a slightly different purpose? It doesn't make it run one, it just makes sure you ran it at least once.

(It could work the way this does, but I rather do that in a follow-up PR.)